### DISCUSSION TITLE
long tables: customization of rule width/thickness

### DISCUSSION THREAD

**Discussion by DAbidou** [lvjr/tabularray#50] [Q&A]
For long tables, it would be great if I could make the width/thickness of the rules encompassing the first header different from those of the repeated headers. Also, the same is desired for the repeated footers and last footer.

**Comment by lvjr** [comment]
You can create your special `tabularray` environment and add extra rules to first head and last foot:

````latex
\documentclass{article}

\usepackage[paperwidth=8cm,paperheight=6cm,width=7cm,height=5cm]{geometry}
\usepackage{tabularray,xcolor}
\UseTblrLibrary{booktabs}

\begin{document}

\DefTblrTemplate{firsthead}{mytemplate}{
  \UseTblrTemplate{caption}{default}
  \vskip 6pt
  \hrule height 1pt
}

\DefTblrTemplate{middlehead,lasthead}{mytemplate}{
  \UseTblrTemplate{capcont}{default}
  \vskip 6pt
}

\DefTblrTemplate{firstfoot,middlefoot}{mytemplate}{
  \vskip 6pt
  \UseTblrTemplate{contfoot}{default}
}

\DefTblrTemplate{lastfoot}{mytemplate}{
  \hrule height 1pt width \tablewidth
  \vskip 6pt
  \UseTblrTemplate{note}{default}
  \UseTblrTemplate{remark}{default}
}

\NewTblrTheme{mytheme}{
  \SetTblrTemplate{head,foot}{mytemplate}
}

\NewTblrEnviron{mytblr}
\SetTblrOuter[mytblr]{long,headsep=0pt,footsep=0pt,theme=mytheme}

\begin{mytblr}[
  caption = {The Caption},
]{
  colspec = {XX}, rowhead = 1,
  row{even} = {gray!30}, row{1} = {blue!15},
}
\toprule[1.5pt]
  One & Two  \\
\midrule
   1  &   2  \\
   3  &   4  \\
   5  &   6  \\
   7  &   8  \\
   9  &  10  \\
  11  &  12  \\
  13  &  14  \\
  15  &  16  \\
  17  &  18  \\
  19  &  20  \\
\bottomrule[1.5pt]
\end{mytblr}

\end{document}
````
![image](https://user-images.githubusercontent.com/4239949/127949030-f80629c0-2118-45c1-8ac8-378c26424022.png)

**Comment by jasperhabicht** [comment]
This solution does not quite work if no lines are printed at the top and bottom of middle parts of the table and `hline`s happen to sit at page breaks. Consider the following example where the line at the bottom of the middle part on page 2 and the line at the top of the middle part on page 3 are missing (also see [question at TeX.SX](https://tex.stackexchange.com/q/681561/47927); probably related to [issue 130](https://github.com/lvjr/tabularray/issues/130)):

```
\documentclass{article}
\usepackage[
    paperwidth=8.5cm, 
    paperheight=5.5cm, 
    width=7.5cm, 
    height=4.5cm]{geometry}
\usepackage{tabularray}

\begin{document}

\DefTblrTemplate{firsthead}{mytemplate}{
  \UseTblrTemplate{caption}{default}
  \vskip 6pt 
  \hrule height 2pt width \tablewidth
  \vskip -6pt 
}

\DefTblrTemplate{lastfoot}{mytemplate}{
  \vskip -2pt 
  \hrule height 2pt width \tablewidth
}

\NewTblrTheme{mytheme}{
  \SetTblrTemplate{firsthead}{mytemplate}
  \SetTblrTemplate{lastfoot}{mytemplate}
}

\begin{longtblr}[
  halign = c,
  caption = {Long Caption},
  label = {tblr:test},
  theme = mytheme,
]{
  colspec = {@{}XXXX@{}},
  hline{3,5,7,9,11} = {1pt},
}
  1 & A & B & C \\
  2 & A & B & C \\
  3 & A & B & C \\
  4 & A & B & C \\
  5 & A & B & C \\
  6 & A & B & C \\
  7 & A & B & C \\
  8 & A & B & C \\
  9 & A & B & C \\
  10 & A & B & C \\ 
  11 & A & B & C \\
  12 & A & B & C \\
\end{longtblr}

\end{document}
```

![image](https://user-images.githubusercontent.com/6378801/229346863-152f3aa6-39b8-4ff6-9e9c-59bc5430da24.png)

**Reply by muzimuzhi** [reply]
> Consider the following example where the line at the bottom of the middle part on page 2 and the line at the top of the middle part on page 3 are missing

They are controlled by `middlefoot` and `lasthead`, respectively, which are both not modified in your example.

**Reply by jasperhabicht** [reply]
Yes, but then there would be `hline`s at every page break which is not wanted. In the above example, on page 2 there should be a line below and on and page 3 there should be a line above the table part, but on page 1 there should be no line below and on page 2 no line above the table part.

**Reply by muzimuzhi** [reply]
Ah... I didn't read the original TeX-SX question until just then.  It seems the OP's request is to not gobble (but only copy) user-specified `hline`s even they appear at page breaks.  Then yes it's not supported by `tabularray`.  Also see https://github.com/lvjr/tabularray/discussions/371#discussioncomment-5115605.

**Reply by muzimuzhi** [reply]
A patch to `\__tblr_build_page_table:nnn` (interaction with `rowhead` and `rowfoot` are not handled):

<details><summary>Full example, new inner options `first hlines` and `last hlines`</summary>
<p>

```tex
\documentclass{article}
\usepackage[
    paperwidth=8.5cm, 
    paperheight=5.5cm, 
    width=7.5cm, 
    height=4.5cm]{geometry}
\usepackage{tabularray}

\ExplSyntaxOn
\bool_new:N \l__tblr_page_table_use_first_hlines_bool
\bool_new:N \l__tblr_page_table_use_last_hlines_bool

\cs_gset_protected:Npn \__tblr_build_page_table:nnn #1 #2 #3
  {
    \__tblr_build_one_table:nnNN {#2} {#3}
      \l__tblr_page_table_use_first_hlines_bool % replaces \c_false_bool
      \l__tblr_page_table_use_last_hlines_bool  % replaces \c_false_bool
    \vbox_set:Nn \l__tblr_table_box
      {
        \box_use:N \l__tblr_table_head_box
        \__tblr_cover_two_vboxes:NN \l__tblr_row_head_box \l__tblr_table_box
        \box_use:N \l__tblr_row_foot_box
        \hrule height ~ 0pt
        \box_use:N \l__tblr_table_foot_box
      }
    \__tblr_halign_whole:Nn \l__tblr_table_box {#1}
  }

% new keys `first/last hlines`
\keys_define:nn { tblr }
  {
    first~ hlines .bool_set:N = \l__tblr_page_table_use_first_hlines_bool,
    last~ hlines  .bool_set:N = \l__tblr_page_table_use_last_hlines_bool
  }
\clist_gput_right:Nn \g__tblr_table_known_keys_clist
  { first~ hlines, last~ hlines }
\ExplSyntaxOff

\begin{document}

\DefTblrTemplate{firsthead}{mytemplate}{
  \UseTblrTemplate{caption}{default}
  \vskip 6pt 
  \hrule height 2pt width \tablewidth
  \vskip -6pt 
}

\DefTblrTemplate{lastfoot}{mytemplate}{
  \vskip -2pt 
  \hrule height 2pt width \tablewidth
}

\NewTblrTheme{mytheme}{
  \SetTblrTemplate{firsthead}{mytemplate}
  \SetTblrTemplate{lastfoot}{mytemplate}
}

\begin{longtblr}[
  halign = c,
  caption = {Long Caption},
  label = {tblr:test},
  theme = mytheme
]{
  colspec = {@{}XXXX@{}},
  hline{3,5,7,9,11} = {1pt},
  first hlines,
  last hlines
}
  1 & A & B & C \\
  2 & A & B & C \\
  3 & A & B & C \\
  4 & A & B & C \\
  5 & A & B & C \\
  6 & A & B & C \\
  7 & A & B & C \\
  8 & A & B & C \\
  9 & A & B & C \\
  10 & A & B & C \\ 
  11 & A & B & C \\
  12 & A & B & C \\
\end{longtblr}

\end{document}
```

</p>
</details>

<details><summary>Full example v2, checked with <code>tabularray</code> 2025B (2025-08-29), options renamed to <code>repeat-first-hlines</code> and <code>repeat-last-hlines</code></summary>
<p>

```tex
\documentclass{article}
\usepackage[
    paperwidth=8.5cm,
    paperheight=5.5cm,
    width=7.5cm,
    height=4.5cm]{geometry}
\usepackage{tabularray}

\usepackage{xpatch}

% checked with tabularray 2025B (2025-08-29)
\ExplSyntaxOn
\bool_new:N \l__tblr_page_repeat_first_hlines_bool
\bool_new:N \l__tblr_page_repeat_last_hlines_bool

% restore catcode of space character. It's needed by the explicit space
% character used in \__tblr_build_page_table:nnn, "\hrule height ~ 0pt"
\char_set_catcode_space:N \
\xpatchcmd \__tblr_build_page_table:nnn
  {%
    \__tblr_build_one_table:nnNN {#2}{#3}\c_false_bool \c_false_bool
  }
  {%
    \__tblr_build_one_table:nnNN {#2}{#3}%
      \l__tblr_page_repeat_first_hlines_bool
      \l__tblr_page_repeat_last_hlines_bool
  }
  {}{\PatchFailed}
\char_set_catcode_ignore:N \

% new inner keys "repeat-first-hlines" and "repeat-last-hlines"
\keys_define:nn { tabularray/table/inner }
  {
    repeat-first-hlines .bool_set:N = \l__tblr_page_repeat_first_hlines_bool,
    repeat-last-hlines  .bool_set:N = \l__tblr_page_repeat_last_hlines_bool
  }
\ExplSyntaxOff

\begin{document}

\DefTblrTemplate{firsthead}{mytemplate}{
  \UseTblrTemplate{caption}{default}
  \vskip 6pt
  \hrule height 2pt width \tablewidth
  \vskip -6pt
}

\DefTblrTemplate{lastfoot}{mytemplate}{
  \vskip -2pt
  \hrule height 2pt width \tablewidth
}

\NewTblrTheme{mytheme}{
  \SetTblrTemplate{firsthead}{mytemplate}
  \SetTblrTemplate{lastfoot}{mytemplate}
}

\begin{longtblr}[
  halign = c,
  caption = {Long Caption},
  label = {tblr:test},
  theme = mytheme
]{
  colspec = {@{}XXXX@{}},
  hline{3,5,7,9,11} = {1pt},
  repeat-first-hlines,
  repeat-last-hlines
}
  1 & A & B & C \\
  2 & A & B & C \\
  3 & A & B & C \\
  4 & A & B & C \\
  5 & A & B & C \\
  6 & A & B & C \\
  7 & A & B & C \\
  8 & A & B & C \\
  9 & A & B & C \\
  10 & A & B & C \\
  11 & A & B & C \\
  12 & A & B & C \\
\end{longtblr}

\end{document}
```

</p>
</details> 

![image](https://user-images.githubusercontent.com/6376638/229377022-f9d72d74-4a61-49f6-b4e7-07438eb7b7de.png)

**Reply by jasperhabicht** [reply]
Nice! This would really be an improvement to the package I think! It is actually not too much code, but to me it was difficult to figure out a solution. Thank you! Maybe you could add this answer to the TeX.SX question as well.

**Reply by ScumCoder** [reply]
@muzimuzhi you are a magician and a savior.

The only thing that separates your code from perfection is that it does not work with `hlines` option; the line at page breaks is repeated whether `first hlines`/`last hlines` are enabled or not.

I tried to cheat by replacing `hlines` with `cell{1-Z}{1} = {cmd=\hline}`, but got a bunch of `You can't use '\hrule' here except with leaders` errors. Looks like I'll have to manually insert `\hline`s everywhere (which is one of the reasons I decided to migrate from `longtable` to `tabularray` in the first place). Oh well.

**Reply by ScumCoder** [reply]
@muzimuzhi OK, I think I found the root of the problem.
The issue with `longtblr`'s hlines is twofold:
1) the package suppresses the typesetting of hlines that sit at page breaks;
2) the package places copies of the hline defined for the top of the very first row and the hline defined for the bottom of the very last row.

Your patch addresses the first issue. Would it be too much to ask you to make a patch that would address the second one too?

**Reply by botanikus** [reply]
Hi there! Has there been an update on this so far?

**Comment by lenoleno** [comment]
Coming from #615.  Following the discussion here, I have managed to put a (non-repeating) horizontal line into `lastfoot`.  But now I'm facing the other problem mentioned by @ScumCoder: 

> the package suppresses the typesetting of hlines that sit at page breaks;

Unfortunately, tabularray 2025A seems to break @muzimuzhi's [patch](https://github.com/lvjr/tabularray/discussions/50#discussioncomment-5503739).  Any idea how to revive that patch?

**Reply by muzimuzhi** [reply]
> Unfortunately, tabularray 2025A seems to break @muzimuzhi's [patch](https://github.com/lvjr/tabularray/discussions/50#discussioncomment-5503739). Any idea how to revive that patch?

Patch updated for `tabularray` 2025A.

**Comment by lenoleno** [comment]
Thanks!  Seems to work fine so far.

One note, since the patch is about *keeping* regular horizontal lines at page breaks, at least for the `repeat-last-hlines` option, I think a name like `keep-bottom-hlines` would be more self-explanatory.  As for the `repeat-first-hlines` option, that one is not required in my use-case, and I really have no idea how hlines at page breaks might interfere with rules above the header.  So I cannot comment on that option name, currently.  Thanks, again!
